### PR TITLE
refactor: introduce typescript

### LINKED ISSUE #119: Introduce TypeScript
It isn't really necessary, but would be nice to have if someone needs a project to do

### PR DESCRIPTION
**Pull request by stickyPiston**
Closes #119

This PR should only be merged once #144 is merged.

### CODE REVIEW

File: src/components/Activities.tsx [hunk 1]
```diff
@@ -1,9 +1,12 @@
 import Activity from './Activity';
 import Poster from './Poster';
-import PropTypes from 'prop-types';
 import { useActivitiesQuery } from '../store/api';
 
-export default function Activities({ current }) {
+type ActivitiesProps = {
```
Comment (josvanos) [inline src/components/Activities.tsx:5]: The state machine props behave like pages or slides. 

I would have created a StateMachineSlideProps kind of interface, which can be reused for quotes, ads and activities. All these 'pages' have the same props

File: src/components/Activity.tsx [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+import moment from 'moment';
+import { type Activity } from '../store/api';
+
+function sameDay(d: Date, t: Date): boolean {
+  return d.getDate() === t.getDate() // getDate returns day number in month...
+    && d.getMonth() === t.getMonth()
+    && d.getFullYear() === t.getFullYear();
+}
+
+function createFormat(has_time: boolean, date: Date, as = new Date()): string {
```
Comment (josvanos) [inline src/components/Activity.tsx:11]: maybe explain what format

File: src/components/Activity.tsx [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+      });
+  }, [active]);
+
+  const startDate = moment(start_date)
+    .format(createFormat(has_start_time, new Date(start_date)));
+  const endDate = end_date
+    ? moment(end_date)
+      .format(createFormat(has_end_time, new Date(end_date), new Date(start_date)))
+    : null;
+  const className = 'activity' + (active ? ' active' : '');
```
Comment (josvanos) [inline src/components/Activity.tsx:40]: It is more readable if you just put it in template  className property.

File: src/components/Activity.tsx [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+  return (
+    <li className={className} ref={activityRef}>
```
Comment (josvanos) [inline src/components/Activity.tsx:43]:
```suggestion
    <li className={`activity ${active ? 'active' : ' '}`} ref={activityRef}>
```

File: src/components/Activity.tsx [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+      <h1>
+        {name}
+        {participant_counter && ` (${participant_counter})`}
```
Comment (josvanos) [inline src/components/Activity.tsx:46]: most of the time in React, you use this syntax for `{condition && <div> </div}` for content and for text a tenerary statement: {participant_counter ? `(${participant_counter})` : ''  }

File: src/snow/snow.ts [hunk 6]
```diff
@@ -5,20 +5,20 @@ const isWinter = (mm >= 12 || mm < 3);
...
-  let link = document.createElement('link');
+  const link = document.createElement('link');
   link.rel = 'stylesheet';
   link.href = 'snow.css';
   document.head.appendChild(link);
-  
+
   // Prepend snow group
-  let snowGroup = document.createElement('div');
+  const snowGroup = document.createElement('div');
```
Comment (josvanos) [inline src/snow/snow.ts:14]: nice improvements

File: src/store/api/contentful.ts [hunk 7]
```diff
@@ -0,0 +1,81 @@
+import { createClient, type EntryFieldTypes, type Entry } from 'contentful';
+import { createApi } from '@reduxjs/toolkit/query/react';
+
+type FromSkeleton<T extends ContentfulSkeletons> = Entry<T, 'WITHOUT_UNRESOLVABLE_LINKS'>['fields'];
+
+type AdSkeleton = {
```
Comment (josvanos) [inline src/store/api/contentful.ts:6]: What do you mean with Skeleton?

File: src/store/api/contentful.ts [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+  fields: {
+    person: EntryFieldTypes.Text,
+    text: EntryFieldTypes.Text
+  }
+};
+type Quote = FromSkeleton<QuoteSkeleton>;
+
+type ContentfulSkeletons = AdSkeleton | BoardMessageSkeleton | QuoteSkeleton;
+
+type ContentfulResponse<ContentType extends ContentfulSkeletons["contentTypeId"]> =
```
Comment (josvanos) [inline src/store/api/contentful.ts:36]: What you are doing here is quite some typescript magic, which makes the code more difficult for beginners. 

I think (/ expect) that use simpler generics, the content is way more easier to understand

File: src/store/api/contentful.ts [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+type BoardMessageSkeleton = {
+  contentTypeId: 'board-message',
+  fields: {
+    message: EntryFieldTypes.Text
+  }
+};
+type BoardMessage = FromSkeleton<BoardMessageSkeleton>;
+
+type QuoteSkeleton = {
+  contentTypeId: 'quotes',
```
Comment (josvanos) [inline src/store/api/contentful.ts:26]: normally in types you specify the type, not the value. 

I think you should create something like this:

```ts
interface ContenfulRequest<T> {
contentTypeId: string;
fields: T;
}

type BoardMessage =  {
message: EntryFieldTypes.Text
}
```

Now you don't need the magic from skeleton function, can type request with `ContenfulRequest<BoardMessage>` and most importantly don't repeat yourself

File: src/store/api/github.ts [hunk 10]
```diff
@@ -1,23 +1,37 @@
 import { Octokit } from 'octokit';
 import { createApi, fakeBaseQuery } from '@reduxjs/toolkit/query/react';
 
-export const octokit = new Octokit({
+const octokit = new Octokit({
   auth: import.meta.env.VITE_GITHUB_API_TOKEN
 });
 
+type Commit = {
```
Comment (josvanos) [inline src/store/api/github.ts:8]: I am more fan of interfaces for objects

File: src/vite-env.d.ts [hunk 11]
```diff
@@ -0,0 +1,21 @@
+/// <reference types="vite/client" />
+
+interface ImportMetaEnv {
```
Comment (josvanos) [inline src/vite-env.d.ts:3]: Nice improvement!

File: tsconfig.json [hunk 12]
```diff
@@ -0,0 +1,92 @@
+{
+  "compilerOptions": {
+    // "incremental": true,                              /* Save .tsbuildinfo files to allow for incremental compilation of projects. */
```
Comment (josvanos) [inline tsconfig.json:3]: please don't specify all options, just show what is not commented out.

**Comment by peppidesu** [comment]
I am blocked by this and 2 other PRs, can I get an ETA on when this will be merged?

**Pull request merged**